From 7d7a3aa2f4f6829720c18592ee93df1674cafca0 Mon Sep 17 00:00:00 2001
From: Roland Haas <roland.haas@physics.gatech.edu>
Date: Wed, 23 Feb 2011 10:14:05 -0500
Subject: [PATCH] expand $1 and $2 in GIT_REPO, DARCS_REPO, HG_REPO

  this patch allows to have a single stanza in a thornlist for a collection of
  git/darcs/hg repositories that share a folder and together make up an
  arrangement. It moves the compuataion of GIT_REPO inside of the checkout loop
  and adds some tests to make sure that the semphores etc. are only initialized
  once.
---
 GetComponents |  144 +++++++++++++++++++++++++++++++--------------------------
 1 files changed, 78 insertions(+), 66 deletions(-)

diff --git a/GetComponents b/GetComponents
index fe27ad3..011aa79 100755
--- a/GetComponents
+++ b/GetComponents
@@ -539,72 +539,6 @@ sub parse_list {
             }
         }
 
-        # parse name of git repo
-        if ( $rec{"TYPE"} eq "git" ) {
-            my $git_repo;
-            # check for custom repo name
-            if (defined($rec{NAME})) {
-                $git_repo = $rec{NAME};
-            } else {
-                $git_repo = $rec{"URL"};
-                $git_repo =~ s/\.git$//;
-                $git_repo =~ s/^.*[:\/]//;
-            }
-            $rec{"GIT_REPO"} = $git_repo;
-            
-            # add the repo to %updated_git_repos and set it to 0
-            # we will use this to track which repos have already been cloned
-            # or updated
-            $updated_git_repos{$git_repo} = 0;
-            if ($PARALLEL) {
-                $semaphores{$git_repo} = Thread::Semaphore->new();
-            }
-        }
-
-        # parse name of darcs repo
-        if ( $rec{"TYPE"} eq "darcs" ) {
-            my $darcs_repo;
-            # check for custom repo name
-            if (defined($rec{NAME})) {
-                $darcs_repo = $rec{NAME};
-            } else {
-                $darcs_repo = $rec{"URL"};
-                $darcs_repo =~ s/_darcs$//;
-                $darcs_repo =~ s/^.*[:\/]//;
-            }
-            $rec{"DARCS_REPO"} = $darcs_repo;
-            
-            # add the repo to %updated_darcs_repos and set it to 0
-            # we will use this to track which repos have already been cloned
-            # or updated
-            $updated_darcs_repos{$darcs_repo} = 0;
-            if ($PARALLEL) {
-                $semaphores{$darcs_repo} = Thread::Semaphore->new();
-            }
-        }
-
-        # parse name of mercurial repo
-        if ( $rec{"TYPE"} eq "hg" ) {
-            my $hg_repo;
-            # check for custom repo name
-            if (defined($rec{NAME})) {
-                $hg_repo = $rec{NAME};
-            } else {
-                $hg_repo = $rec{"URL"};
-                $hg_repo =~ s/\.hg$//;
-                $hg_repo =~ s/^.*[:\/]//;
-            }
-            $rec{"HG_REPO"} = $hg_repo;
-            
-            # add the repo to %updated_hg_repos and set it to 0
-            # we will use this to track which repos have already been cloned
-            # or updated
-            $updated_hg_repos{$hg_repo} = 0;
-            if ($PARALLEL) {
-                $semaphores{$hg_repo} = Thread::Semaphore->new();
-            }
-        }
-
         # save target in original form to check existence.
         my $target = $rec{"TARGET"};
 
@@ -642,6 +576,84 @@ sub parse_list {
                 $rec{AUTH_URL} = $rec{"AUTH_URL_ORIG"};
                 $rec{AUTH_URL} =~ s!\$1!$dir1!;
                 $rec{AUTH_URL} =~ s!\$2!$dir2!;
+            }  
+
+            # parse name of git repo
+            if ( $rec{"TYPE"} eq "git" ) {
+                my $git_repo;
+                # check for custom repo name
+                if (defined($rec{NAME})) {
+                    $git_repo = $rec{NAME};
+                } else {
+                    $git_repo = $rec{"URL"};
+                    $git_repo =~ s/\.git$//;
+                    $git_repo =~ s/^.*[:\/]//;
+                }
+                $rec{"GIT_REPO"} = $git_repo;
+                $rec{"GIT_REPO"} =~ s!\$1!$dir1!;
+                $rec{"GIT_REPO"} =~ s!\$2!$dir2!;
+                
+                # add the repo to %updated_git_repos and set it to 0
+                # we will use this to track which repos have already been cloned
+                # or updated
+                if (not exists $updated_git_repos{$git_repo}) {
+                    $updated_git_repos{$git_repo} = 0;
+                }
+                if ($PARALLEL and not exists $semaphores{$git_repo}) {
+                    $semaphores{$git_repo} = Thread::Semaphore->new();
+                }
+            }
+    
+            # parse name of darcs repo
+            if ( $rec{"TYPE"} eq "darcs" ) {
+                my $darcs_repo;
+                # check for custom repo name
+                if (defined($rec{NAME})) {
+                    $darcs_repo = $rec{NAME};
+                } else {
+                    $darcs_repo = $rec{"URL"};
+                    $darcs_repo =~ s/_darcs$//;
+                    $darcs_repo =~ s/^.*[:\/]//;
+                }
+                $rec{"DARCS_REPO"} = $darcs_repo;
+                $rec{"DARCS_REPO"} =~ s!\$1!$dir1!;
+                $rec{"DARCS_REPO"} =~ s!\$2!$dir2!;
+                
+                # add the repo to %updated_darcs_repos and set it to 0
+                # we will use this to track which repos have already been cloned
+                # or updated
+                if (not exists $updated_darcs_repos{$darcs_repo}) {
+                    $updated_darcs_repos{$darcs_repo} = 0;
+                }
+                if ($PARALLEL and not exists $semaphores{$darcs_repo}) {
+                    $semaphores{$darcs_repo} = Thread::Semaphore->new();
+                }
+            }
+    
+            # parse name of mercurial repo
+            if ( $rec{"TYPE"} eq "hg" ) {
+                my $hg_repo;
+                # check for custom repo name
+                if (defined($rec{NAME})) {
+                    $hg_repo = $rec{NAME};
+                } else {
+                    $hg_repo = $rec{"URL"};
+                    $hg_repo =~ s/\.hg$//;
+                    $hg_repo =~ s/^.*[:\/]//;
+                }
+                $rec{"HG_REPO"} = $hg_repo;
+                $rec{"HG_REPO"} =~ s!\$1!$dir1!;
+                $rec{"HG_REPO"} =~ s!\$2!$dir2!;
+                
+                # add the repo to %updated_hg_repos and set it to 0
+                # we will use this to track which repos have already been cloned
+                # or updated
+                if (not exists $updated_hg_repos{$hg_repo}) {
+                    $updated_hg_repos{$hg_repo} = 0;
+                }
+                if ($PARALLEL and not exists $semaphores{$hg_repo}) {
+                    $semaphores{$hg_repo} = Thread::Semaphore->new();
+                }
             }
 
             $rec{"CHECKOUT"} = $checkout;
-- 
1.7.1

